### PR TITLE
feat: support filter in flat run table

### PR DESCRIPTION
**Pull request by keita-determined**
## Ticket

[ET-111]

## Description

Add filter feature in Flat Run table (Run table is under feature flag)

## Test Plan

- Enable feature flag of Flat Runs View
- Go to flat run table
- Check if 
  - filter button is visible and usable on top of the run table
  - filter is saved after applying valid filters, so filtered table is preserved even after refreshing the page
  - filter should be able to be applied from table header
  - filter should be able to be cleared from table header
(filter behavior should be the same as the experiment table we have now (its not feature flagged))

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[ET-111]: https://hpe-aiatscale.atlassian.net/browse/ET-111?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/pages/FlatRuns/FlatRuns.tsx [hunk 1]
```diff
@@ -606,21 +648,77 @@ const FlatRuns: React.FC<Props> = ({ project }) => {
...
+          label: 'Hide column',
+          onClick: () => {
+            const newColumnsOrder = columnsIfLoaded.filter((c) => c !== column?.column);
+            handleColumnsOrderChange?.(newColumnsOrder);
+            if (isPinned) {
+              handlePinnedColumnsCountChange?.(Math.max(settings.pinnedColumnsCount - 1, 0));
+            }
+          },
+        },
       ];
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRuns.tsx]: removed

File: webui/react/src/pages/FlatRuns/FlatRuns.tsx [hunk 2]
```diff
@@ -424,19 +446,38 @@ const FlatRuns: React.FC<Props> = ({ project }) => {
...
+              });
+            } else {
+              formStore.init(formSetValidation.right);
+            }
+          }
+          cleanup = formStore.asJsonString.subscribe(() => {
+            resetPagination();
+            const loadableFormset = formStore.formset.get();
+            Loadable.forEach(loadableFormset, (formSet) =>
+              updateSettings({ filterset: JSON.stringify(formSet) }),
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRuns.tsx]: done

**Review by johnkim-det — COMMENTED** [review]
- noticed that Searcher Metrics Value does not show filter options in header menu but it does still appear in the Filters dropdown menu. This is true for Experiments List, too. But maybe worth addressing here?

- Also, it looks like the Searcher ID and Searcher Description columns appear in the table but can't be filtered, even though they're not in `BANNED_FILTER_COLUMNS` (and I don't think we would want them to be?). I think we just need to make sure everything in `getColumnDefs` and `runColumns` matches. May need to update again with #9146 but should probably make sure current columns work with filters as expected.

- Also not sure why, but I don't think ID filter is working. Returns no results even when I'm expecting one to be returned.

File: webui/react/src/pages/FlatRuns/FlatRuns.tsx [hunk 3]
```diff
@@ -607,20 +649,64 @@ const FlatRuns: React.FC<Props> = ({ project }) => {
...
         items.push(
-          ...(BANNED_FILTER_COLUMNS.includes(column.column)
-            ? []
-            : [
-                { type: 'divider' as const },
-                ...sortMenuItemsForColumn(column, sorts, handleSortChange),
-              ]),
+          { type: 'divider' as const },
+          ...sortMenuItemsForColumn(column, sorts, handleSortChange),
+          { type: 'divider' as const },
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRuns.tsx:689]: fixed

**Comment by keita-determined** [comment]
> - noticed that Searcher Metrics Value does not show filter options in header menu but it does still appear in the Filters dropdown menu. This is true for Experiments List, too. But maybe worth addressing here?

fixed in run table and experiment table

> - Also, it looks like the Searcher ID and Searcher Description columns appear in the table but can't be filtered, even though they're not in BANNED_FILTER_COLUMNS (and I don't think we would want them to be?). I think we just need to make sure everything in getColumnDefs and runColumns matches. May need to update again with https://github.com/determined-ai/determined/pull/9146 but should probably make sure current columns work with filters as expected.

`Searcher ID` and `Searcher Description` filters exit. 
in https://github.com/determined-ai/determined/blob/e31135ae65352c776df97e58335ac133b5d37fe0/webui/react/src/pages/FlatRuns/columns.ts#L236, but the column names are overwritten, so `Searcher ID` filter exits as `ID`. `Search Name` filter exits as `Name`. 
@johnkim-det what do you think is the best solution for this? If we wanna keep the current column names, we need to fix backend. if we wanna change the column name, these should match the column data from backend. i think backend change is required for either way


> - Also not sure why, but I don't think ID filter is working. Returns no results even when I'm expecting one to be returned.

`ID` means `run id` instead of `experiment id`? looks like backend doesn't support it yet. I don't think frontend can do much with it afaik.

**Comment by keita-determined** [comment]
> But I'm less concerned with the mismatch of column names between filters menu and column headers than the fact that the column header menus aren't displaying filter options at all for those columns. the name mismatch shouldn't cause that, right?

Right, the name mismatch doesnt cause the issue. Backend doesnt return these columns. thats the issue. i dont think web cant do anything with this until backend return missing columns

**Review by johnkim-det — APPROVED** [review]
OK I see now that the filter options are based on `projectColumns` while displaying the column is based on `settings.columns`/FE definition. Not sure if it's good to have two different places where columns are defined like that, but I guess it's out of scope of this PR either way.

It does look like #9146 will provide `experimentId` and `experimentDescription` in `projectColumns`. might be a good idea to try those changes with yours to make sure that at least everything filter-related will work.

but will approve this assuming those two columns are the only issues, since they'd be addressed with #9146 merge.

ID filter not working is a different problem, do you mind making a ticket for that so it's captured for grooming discussion?

**Comment by keita-determined** [comment]
made [a ticket](https://hpe-aiatscale.atlassian.net/browse/ET-253)

**Pull request merged**